### ISSUE TITLE
Improved LLM usage reporting

### ISSUE THREAD

**Issue by nsobadzhiev** [stanfordnlp/dspy#685]
LLMs are expensive and especially with excessive use of Multi-hop steps and Chains of Thought, the token count and number of completions can get out of hand.

I believe that this is on many developers' mind and they would appreciate additional tooling in regards to what their pipelines are doing.

# Current situation

As far as I see, the easiest way to diagnose what DSPy used an LLM for, is to use the `LM` subclass. There, `inspect_history` can be used in order to print the prompts and completions that happened.
Alternatively, `LM` exposes the `history` list that contains more information about the LLM requests that ran.

# The proposal

## A minimal change

A small change to the codebase could be to add another method, similar to `inspect_history` that focuses on cost-related fields (e.g. the usage sub-dict containing the number of token in the prompt and completion):

```python
def summarize_usage(history_item: dict[str, any]):
    usage = history_item['response']['usage']
    print(f"LLM usage: {usage['prompt_tokens']} prompt, {usage['completion_tokens']} completion, {usage['total_tokens']} total")
```

A nice to have would be to return these values as a dict, or a domain object, instead of just printing them. That way, developers can fetch that data and use it for metrics, diagnostics etc.

## A more sophisticated solution

We could think about a broader solution that integrates these metrics deeper into the library's lifecycle. What would for sure be nice, would be to bundle related history list items so it's easier to understand which `usage` was for which prediction.

Apart from that, I'm drawing a blank as I'm still not very familiar with the codebase.

I would be happy to introduce these changes, or whatever might be better than my proposal. I just need feedback from the team if that would be a desirable change and if this approach would be appropriate.

**Comment by KCaverly** [comment]
This is likely going to be possible on a Language Model specific basis. For example, LiteLLM makes costing super simple, and this would be very simple to implement. With this backend refactor we are working on, we track much more detailed LM outputs, and it would be pretty trivial to introduce costing to our lm history tracking if the model provides this information. 

The comment you make surrounding which prediction is generating which usage has been brought up a few times as well. I don't think we have a great way to track which Prediction class/program is calling the LM, this likely needs deeper exploration, but would be super valuable.

**Comment by KCaverly** [comment]
If you're keen you are welcome to take a look at the new backends infra [here](https://github.com/stanfordnlp/dspy/tree/backend-refactor/dspy/backends). We've been working on it in the backend-refactor branch. Currently we've just got LiteLLM as a backend, which manages for the majority of LMs we currently have in dsp.modules.

I think the first three questions we would want to iron out are:
1. What information do we want to store on a per call basis to track usage/cost.
    - We already have an `LMOutput` object, which we could extend to include metadata about usage.
2. How could we link the LM generation itself to a `dspy.Module`?
3. How could we expose this data so its useful to users?

**Comment by nsobadzhiev** [comment]
Alright! Let me look at the way things will be working in the future and see if I can come up with a nice way to expose some metrics there.
What I'm still missing a bit is the structure of each supported LLM's output. I've only worked with a handful of LLMs. I know LiteLLM unifies many by translating them to the API openAI uses, but you probably support LLMs that cannot be used with LiteLLM?

In any case, I'll come back to you and team team if I have an idea how to proceed.
What should we do with this Issue? Should it remain open or should we close it and see if we can open something more concrete later on?

**Comment by KCaverly** [comment]
We should have the flexibility to handle any LM that implements the BaseLM class. As we can’t guarantee that every LM will offer costing on a per request basis, maybe the output metadata associated with costing/usage should be optional?

**Comment by nsobadzhiev** [comment]
Here's my take, based on what I'm seeing in the backend refactor PR.

Pydantic models are being used already relatively extensively, which is cool - I personally like that. And there's already the `LMOutput` class we can piggyback onto. I would another field like `metrics` or `stats` or `costs` to it. That field would be of a new `BaseModel` subclass.

### TL;DR - my proposal

* I'd add another BaseModel for cost metrics and add it as an optional field to the `LMOutput`.
* I'd let each `BaseLM` subclass populate it
* I'd add a base implementation that just counts/estimates the tokens in the prompt and the output from the LM

```python
class BaseMetrics(BaseModel):
    prompt_tokens: int,
    completion_tokens: int,
    total_tokens: int,
    model: str,   # maybe not necessary
```

### Handling different LLMs

As discussed, not all LLMs expose these cost metrics, or might have them differently. I agree that this new field would have to be optional.
In the base implementation in `BaseLM` we could leave it empty and let each subclass optionally parse it as part of the `__call__` or `generate` method. Obviously LiteLM would parse the openAI style cost json field.

### Default cost metrics

One thought I had was - for a base set of metrics, do we actually need an LLM to report anything in the response? In The openAI case, for, instance, the most useful data you get is how many tokens you used. But this is information you would infer yourself by calculating or estimating the number of tokens in the prompt and in the completion.

So can't we have a base implementation that fills these cost metrics purely by calculating the size of the input and output of the request. If we do this on the `BaseLM` level, we capture the changes to the prompt and each intermediate step (e.g. a Chain of thought, re-phrasing a question, adding context to the prompt) which, for me, is the most interesting thing - how much bigger and costlier do my prompts get by using DSPy magic.

### Grouping cost metrics by pipeline

Now I understand why you said it would be difficult to know which usage was incurred by which prediction. Based on what I was in the PR, I see that such logic would have to be on a higher level than `BaseLM`. I see there's the `BaseBackend` classes, which I don't completely understand, since they are still not used anywhere. My gut feeling is that they are still too low-level to be able to group completions into prediction runs, but maybe there will be a layer on top of them what encompasses all completions for a run of the pipeline, and there we could just run through all requests (much like the `history` array in `BaseLM`)?

**Comment by nsobadzhiev** [comment]
> What is best workaround right now ?

Unless something changed in the meantime, I believe the `inspect_history` function on `LM` classes is still your best bet. It will print out and return the last N prompts and completions. As far as I remember, the output highly depends on the LLM used, but that's alright unless you need a high degree of flexibility. You might even consider creating your own LLM subclass. In any case, the struggle here is that it's not trivial to know who invoked these last N requests and also what N should be (no way of knowing how many requests were made since you wanted inspected the history)

Last time I checked, there were also some libraries that collect LLM metrics that could help. Looking at their code, it seemed to me that they were wrapping around the LLM client code (the OpenAI library for instance) and extracting the usage every time someone requested a completion. That might help you but it's not part of DSPy. You might also have to create your own LM subclass for this to work.
I don't remember which library I saw this one in. A quick search today yielded [langkit](https://github.com/whylabs/langkit), but I'm not sure.

**Comment by okhat** [comment]
This is now addressed in the new dspy.LM class

**Comment by lorenzobalzani** [comment]
> This is now addressed in the new dspy.LM class

Hi, where exactly?

**Comment by RemoSande** [comment]
@okhat Hi, how exactly is this addressed by the LM class ? 

DSPy is fantastic but without token tracking it is like playing darts in the dark : we need to know token-usage / pricing / relevant metrics to accurately asses our results.

**Comment by cgint** [comment]
Maybe in case someone stumbles over this one like i did ...

The following works for me using dspy 3.0.0b3 and outputs
```json
{
  "vertex_ai/gemini-2.5-flash": {
    "completion_tokens": 2945,
    "prompt_tokens": 9456,
    "total_tokens": 12401,
    "completion_tokens_details": {
      "accepted_prediction_tokens": 0,
      "audio_tokens": 0,
      "reasoning_tokens": 1782,
      "rejected_prediction_tokens": 0,
      "text_tokens": 1163
    },
    "prompt_tokens_details": {
      "audio_tokens": 0,
      "cached_tokens": 0,
      "text_tokens": 5328,
      "image_tokens": 0
    }
  }
}
```

Main points:
- Set track_usage=True in dspy.settings.configure
- Use output.get_lm_usage()

```python
import dspy

# Configure DSPy with LM and usage tracking enabled
dspy.settings.configure(
    lm=dspy.LM("openai/gpt-4o-mini", cache=False),
    track_usage=True
)

# Define a simple DSPy module that calls the LM twice
class MyProgram(dspy.Module):
[...]

# Instantiate and run the program
program = MyProgram()
output = program(question="What is the capital of France?")

# Get LM usage statistics
print(output.get_lm_usage())

```
